google/cloud/bigquery/v2/minimal/internal/dataset.h

```diff
@@ -0,0 +1,195 @@
+
+struct TableReference {
+  std::string project_id;
+  std::string dataset_id;
+  std::string table_id;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(TableReference, dataset_id,
+                                                project_id, table_id);
+
+struct RoutineReference {
+  std::string project_id;
+  std::string dataset_id;
+  std::string routine_id;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(RoutineReference, dataset_id,
+                                                project_id, routine_id);
+
+struct DatasetAccessEntry {
+  DatasetReference dataset;
+  std::vector<TargetType> target_types;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(DatasetAccessEntry, dataset,
+                                                target_types);
+
+struct Access {
+  bool operator==(Access const& other);
+
+  std::string role;
+  std::string user_by_email;
+  std::string group_by_email;
+  std::string domain;
+  std::string special_group;
+  std::string iam_member;
+
+  TableReference view;
+  RoutineReference routine;
+  DatasetAccessEntry dataset;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(Access, role, user_by_email,
+                                                group_by_email, domain,
+                                                special_group, iam_member, view,
+                                                routine, dataset);
+
+struct HiveMetastoreConnectivity {
+  std::string access_uri_type;
+  std::string access_uri;
+  std::string metadata_connection;
+  std::string storage_connection;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(HiveMetastoreConnectivity,
+                                                access_uri_type, access_uri,
+                                                metadata_connection,
+                                                storage_connection);
+
+struct HiveDatabaseReference {
+  std::string catalog_id;
+  std::string database;
+
+  HiveMetastoreConnectivity metadata_connectivity;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(HiveDatabaseReference,
+                                                catalog_id, database,
+                                                metadata_connectivity);
+
+struct ExternalDatasetReference {
+  HiveDatabaseReference hive_database;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(ExternalDatasetReference,
+                                                hive_database);
+
+struct GcpTag {
+  std::string tag_key;
+  std::string tag_value;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(GcpTag, tag_key, tag_value);
+
+struct Dataset {
+  std::string kind;
+  std::string etag;
+  std::string id;
+  std::string self_link;
+  std::string friendly_name;
+  std::string description;
+  std::string type;
+  std::string location;
+  std::string default_collation;
+
+  bool published = false;
+  bool is_case_insensitive = false;
+
+  std::chrono::system_clock::time_point default_table_expiration_ms =
+      kDefaultTimepoint;
```

Comment:
This is probably wrong.  If the time is `...::timepoint` then the units should not be in the name (`..._ms`).  And you are just writing a complicated expression to calll the default constructor. This would be equivalent:

```cc
  std::chrono::system_clock::time_point default_partition_expiration;
```